### PR TITLE
Data loading

### PR DESCRIPTION
**Pull request by LouGel**
certifier & projects-sdg loading

### CODE REVIEW

File: src/contribution-manager/certifier/certifier.controller.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import { CertifierService } from './certifier.service';
+import { Roles } from '../../roles/roles.decorator';
+import { Role } from '../../roles/role.enum';
+
+@ApiTags('certifier')
+@Controller('certifier')
+export class CertifierController {
+  private readonly logger = new Logger(CertifierController.name);
+
+  constructor(private companyService: CertifierService) {}
```
Comment (MartianGreed) [inline src/contribution-manager/certifier/certifier.controller.ts:19]: `companyService` or  `certifierService`  👀 ?

File: src/contribution-manager/certifier/certifier.controller.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  @UseInterceptors(FileInterceptor('file'))
+  @ApiOperation({ summary: 'Upload certifier CSV file' })
+  @ApiResponse({ status: 201, description: 'File successfully processed.' })
+  @ApiResponse({ status: 400, description: 'Invalid file format.' })
+  async uploadCertifierCSV(
+    @UploadedFile() file: Express.Multer.File,
+  ): Promise<{ message: string }> {
+    this.logger.debug(`File uploaded: ${file.originalname}`);
+
+    return await this.companyService.processCsv(file.buffer);
```
Comment (MartianGreed) [inline src/contribution-manager/certifier/certifier.controller.ts:32]: same here

File: src/contribution-manager/project-sdgs/project-sdgs.controller.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import { ProjectSdgsService } from './project-sdgs.service';
+import { Roles } from '../../roles/roles.decorator';
+import { Role } from '../../roles/role.enum';
+
+@ApiTags('project-sdgs')
+@Controller('project-sdgs')
+export class ProjectSdgsController {
+  private readonly logger = new Logger(ProjectSdgsController.name);
+
+  constructor(private companyService: ProjectSdgsService) {}
```
Comment (MartianGreed) [inline src/contribution-manager/project-sdgs/project-sdgs.controller.ts:19]: var naming

File: src/contribution-manager/project-sdgs/project-sdgs.controller.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  @UseInterceptors(FileInterceptor('file'))
+  @ApiOperation({ summary: 'Upload project_sdgs CSV file' })
+  @ApiResponse({ status: 201, description: 'File successfully processed.' })
+  @ApiResponse({ status: 400, description: 'Invalid file format.' })
+  async uploadProjectSdgsCSV(
+    @UploadedFile() file: Express.Multer.File,
+  ): Promise<{ message: string }> {
+    this.logger.debug(`File uploaded: ${file.originalname}`);
+
+    return await this.companyService.processCsv(file.buffer);
```
Comment (MartianGreed) [inline src/contribution-manager/project-sdgs/project-sdgs.controller.ts:31]: 👀

File: src/csv/csv.service.ts [hunk 5]
```diff
@@ -1,9 +1,23 @@
...
+    return parsed;
+  };
+
+  parseJSONSafe = (value: string): Prisma.JsonValue => {
+    try {
+      return JSON.parse(value);
+    } catch (error) {
+      throw new Error(`Invalid JSON: ${value}`);
+    }
+  };
```
Comment (MartianGreed) [inline src/csv/csv.service.ts:20]: functions can be defined as private

File: src/csv/csv.service.ts [hunk 6]
```diff
@@ -12,13 +26,15 @@ export class CsvService {
...
+        .pipe(csv({ strict: true }))
+        .on('data', (data) => {
+          results.push(data as T);
+        })
         .on('end', () => {
           resolve(results);
         })
         .on('error', (error) => {
-          reject(error);
+          reject(new BadRequestException(`${error}`));
```
Comment (MartianGreed) [inline src/csv/csv.service.ts:37]: not sure if this service is responsible of throwing that kind of error : 
IMO : services throws `InvalidInputData` + some details about line in file (maybe `error` already carries this information)
controller catches that exception to throw back a `BadRequestException`

File: src/feature/specs/add-allocation.steps.ts [hunk 7]
```diff
@@ -161,6 +161,7 @@ defineFeature(feature, (test) => {
       const orders = await orderBookRepository.findByBusinessUnitIds([
         businessId,
       ]);
+      void orderCount;
```
Comment (MartianGreed) [inline src/feature/specs/add-allocation.steps.ts:164]: we should leave the error as long as test is not valid (hardcoded value)

**Review by MartianGreed — CHANGES_REQUESTED** [review]
Overall good ! 

For maintanability, `CsvService` could handle all of the `business-unit`, `company` and `certifier` services. 
I think we need to add this behaviour so that we won't have to create more services, just configuration passed to `CsvService`

File: src/csv/csv.service.ts (hunk 5 above)
Comment (LouGel) [reply src/csv/csv.service.ts:20]: Needeb buy businessunits service

File: src/csv/csv.service.ts (hunk 6 above)
Comment (LouGel) [reply src/csv/csv.service.ts:37]: The pris error, code and message are sometimes shady and say Unknow error. Wanted to wrap it to be clearer.

**Review by julienbrs — APPROVED** [review]

**Pull request merged**